# Exercise 4.8 Grouping and Transforming Data

Import Libraries

In [1]:
# Import libraries 
import pandas as pd
import numpy as np
import os

Import Dataset

In [2]:
# Create path shortcut
path = r'C:\Users\kacee\OneDrive\Desktop\Instacart Basket Analysis'

In [3]:
# Import pkl orders_products_merged_groupby
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_groupby.pkl'))

In [4]:
# Check output of imported dataframe
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

Aggregate mean of order_number grouped by department_id

In [5]:
# Average number of orders per user across each Instacart department
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Analysis of the result:

In the subset df there were only 8 departments that appeared in the results out of the full 21 that appeared from the results of the entire dataframe. The subset had a minimum mean of 11.29 and a maximum mean of 19.46. Results from the entire dataframe show a more accurate representation with a minimum mean of 15.21 and a maximum mean of 22.9.

Create a loyalty flag for existing customers using the transform() and loc() functions

In [6]:
# New column- Max order per customer
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
# Check output with max order column
df_ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda      

In [8]:
# Create a loyalty flag
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [9]:
df_ords_prods_merge

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc     busiest_day  \
0               77              7     9.0  Mid-range product  Regularly busy   
1               77              7     9.0  Mid-range product  Regularly busy   
2               77              7     9.0  Mid-range product  Regularly busy   
3               77              7     9.0  Mid-range product      Least busy   
4               77              7     9.0  Mid-range product      Least busy   
...            ...            ...     ...                ...             ...   
32404854        64              7     3.7  Low-range product  Regularly busy   
32404855        64              7     3.7  Low-range product  Regularly busy   
32404856        50             19     6.9  Mid-range product  Regularly busy   
32404857        50             19     6.9  Mid-range product  Regularly busy   
32404858       130             14    10.3  Mid-range product      Least busy   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
0          Regularly busy        Average orders         10      New customer  
1         Least busy days        Average orders         10      New customer  
2         Least busy days           Most orders         10      New customer  
3         Least busy days        Average orders         10      New customer  
4         Least busy days           Most orders         10      New customer  
...                   ...                   ...        ...               ...  
32404854   Regularly busy           Most orde

In [10]:
# Count loyalty_flag values
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

Use the loyalty flag to check the basic statistics of product prices for each loyalty category 

In [11]:
# Group data by user_id, find mean product price for each user. mean_price for results
df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [12]:
# Check output
df_ords_prods_merge[['user_id', 'mean_price']].head(20)

user_id  mean_price
0         1    6.367797
1         1    6.367797
2         1    6.367797
3         1    6.367797
4         1    6.367797
5         1    6.367797
6         1    6.367797
7         1    6.367797
8         1    6.367797
9         1    6.367797
10       15    3.980556
11       15    3.980556
12       15    3.980556
13       15    3.980556
14       15    3.980556
15       19    8.372059
16       19    8.372059
17       19    8.372059
18       21    7.012683
19       31    8.137793

In [13]:
# Check output
df_ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda      

In [14]:
# Check output
df_ords_prods_merge

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc     busiest_day  \
0               77              7     9.0  Mid-range product  Regularly busy   
1               77              7     9.0  Mid-range product  Regularly busy   
2               77              7     9.0  Mid-range product  Regularly busy   
3               77              7     9.0  Mid-range product      Least busy   
4               77              7     9.0  Mid-range product      Least busy   
...            ...            ...     ...                ...             ...   
32404854        64              7     3.7  Low-range product  Regularly busy   
32404855        64              7     3.7  Low-range product  Regularly busy   
32404856        50             19     6.9  Mid-range product  Regularly busy   
32404857        50             19     6.9  Mid-range product  Regularly busy   
32404858       130             14    10.3  Mid-range product      Least busy   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  \
0          Regularly busy        Average orders         10      New customer   
1         Least busy days        Average orders         10      New customer   
2         Least busy days           Most orders         10      New customer   
3         Least busy days        Average orders         10      New customer   
4         Least busy days           Most orders         10      New customer   
...                   ...                   ...        ...               ...   
32404854   Regularly busy           Mo

In [15]:
# Check basic statistics for each customer loyalty group
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Create a spending flag for each user based on average price across all their orders 

In [16]:
# Create spending flags to user ID based on mean product price
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [17]:
# Count spending_flag values
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [18]:
df_ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   

   product_name  ...  department_id  prices    price_range_loc  \
0          Soda  ...              7     9.0  Mid-range product   
1          Soda  ...              7     9.0  Mid-range product   
2          Soda  ...              7     9.0  Mid-range product   
3          Soda  ...              7     9.0  Mid-range product   
4          Soda  ...              7     9.0  Mid-range product   
5          Soda  ...              7     9.0  Mid-range product   
6          Soda  ...              7     9.0  Mid-range product   
7          Soda  ...              7     9.0  Mid-range product   
8          Soda  ...              7     9.0  Mid-range product   
9          Soda  ...              7     9.0  Mid-range product   
10         Soda  ...              7     9.0  Mid-range product   
11         Soda  ...              7     9.0  Mid-range product   
12         Soda

Determine frequent versus non-frequent customers

In [19]:
# Group data by user_id, find median product of days since last order. median_days for results
df_ords_prods_merge['median_days'] = df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [20]:
# Check output
df_ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   
15                    6.0         196                  3          0   
16                    8.0         196                 15          1   
17                    6.0         196                  1          1   
18                    6.0         196                  2          0   
19                    7.0         196                 17          0   

   product_name  ...  prices    price_range_loc     busiest_day  \
0          Soda  ...     9.0  Mid-range product  Regularly busy   
1          Soda  ...     9.0  Mid-range product  Regularly busy   
2          Soda  ...     9.0  Mid-range product  Regularly busy   
3          Soda  ...     9.0  Mid-range product      Least busy   
4          Soda  ...     9.0  Mid-range product      Least busy   
5          Soda  ...     9.0  Mid-range product  Regularly busy   
6          Soda  ...     9.0  Mid-range product  Regularly busy   
7          Soda  ...     9.0  Mid-range product  Regularly busy   
8          Soda  ...     9.0  Mid-range product  Regularly busy   
9          Soda  ...     9.0  Mid-range product      Least busy   
10         Soda  ...     9.0  Mid-range product  Regularly busy   
11         Soda  ...     9.0  Mid-range product  Regularly busy   
12

In [21]:
# Create frequency flags per user_id based on median days since prior order
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days'] > 10) & (df_ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [22]:
# Count frequency flags
df_ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

Export Dataset

In [23]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_aggregated.pkl'))